File: Perl/abuild-1.1.6/src/QTC.cc
#include <QTC.hh>

#include <boost/thread/mutex.hpp>
#include <string>
#include <set>
#include <stdio.h>
#include <stdlib.h>
#include <Util.hh>
#ifdef _WIN32
#include <windows.h>
#endif

// With some compilers (gcc 4.1.0 on Fedora Core 5), declaring these
// statics inside QTC::TC causes problems when they are accessed by
// multiple threads, but declaring then static at the file level works
// fine.
static std::set<std::pair<std::string, int> > cache;
static boost::mutex qtc_mutex;

static bool tc_active(char const* const scope)
{
    std::string value;
    return (Util::getEnv("TC_SCOPE", &value) && (value == scope));
}

void QTC::TC(char const* const scope, char const* const ccase, int n)
{
    if (! tc_active(scope))
    {
	return;
    }

    std::string filename;
#ifdef _WIN32
# define TC_ENV "TC_WIN_FILENAME"
#else
# define TC_ENV "TC_FILENAME"
#endif
    if (! Util::getEnv(TC_ENV, &filename))
    {
	return;
    }
#undef TC_ENV

    boost::mutex::scoped_lock lock(qtc_mutex);
    if (cache.count(std::make_pair(ccase, n)))
    {
	return;
    }
    cache.insert(std::make_pair(ccase, n));

    FILE* tc = fopen(filename.c_str(), "ab");
    if (tc)
    {
	fprintf(tc, "%s %d\n", ccase, n);
	fclose(tc);
    }
}

File: Perl/abuild-1.1.6/src/java-support/src/java/org/abuild/QTC.java
package org.abuild;

import java.util.HashSet;
import java.util.Set;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.io.IOException;

public class QTC
{
    private static String filename = getFilename();
    private static String tc_scope = System.getenv("TC_SCOPE");
    private static Set<String> cache = new HashSet<String>();

    private static String getFilename()
    {
	String filename = System.getenv("TC_WIN_FILENAME");
	if (filename == null)
	{
	    filename = System.getenv("TC_FILENAME");
	}
	return filename;
    }

    public static void TC(String scope, String ccase)
    {
	TC(scope, ccase, 0);
    }

    synchronized public static void TC(String scope, String ccase, int n)
    {
	if (! tcActive(scope))
	{
	    return;
	}

	String key = ccase + " " + n;
	if (cache.contains(key))
	{
	    return;
	}
	cache.add(key);

	try
	{
	    PrintWriter tc = new PrintWriter(new FileWriter(filename, true));
	    tc.println(key);
	    tc.close();
	}
	catch (IOException e)
	{
	    // ignore
	}
    }

    static private boolean tcActive(String scope)
    {
	return ((filename != null) &&
		(tc_scope != null) &&
		(tc_scope.equals(scope)));
    }
}
